File: src/main/java/com/bdev/hogwarts_api/rest_service/student_status/StudentStatusRestServiceImpl.java
package com.bdev.hogwarts_api.rest_service.student_status;

import acropollis.municipali.security.common.dto.MunicipaliUserInfo;
import com.bdev.hogwarts_api.data.dto.student.Student;
import com.bdev.hogwarts_api.data.dto.student.StudentStatus;
import com.bdev.hogwarts_api.data.dto.student.StudentStatusType;
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException;
import com.bdev.hogwarts_api.service.student.StudentService;
import com.bdev.hogwarts_api.service.student_status.StudentStatusService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class StudentStatusRestServiceImpl implements StudentStatusRestService {
    @Autowired
    private StudentService studentService;
    @Autowired
    private StudentStatusService studentStatusService;

    @Transactional(readOnly = true)
    @Override
    public List<StudentStatus> getLatestStudentsStatuses(MunicipaliUserInfo userInfo) {
        return studentService
                .getAllStudents()
                .stream()
                .map(it -> studentStatusService.getStudentStatus(it.getId()))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    @Override
    public List<StudentStatus> getStudentStatuses(MunicipaliUserInfo userInfo, long studentId) {
        if (!studentService.exists(studentId)) {
            throw new HttpEntityNotFoundException("Student with id '%d' does not exist", studentId);
        }

        return studentStatusService.getAllStudentStatuses(studentId);
    }

    @Transactional
    @Override
    public void changeStudentStatus(MunicipaliUserInfo userInfo, long studentId, StudentStatusType status, long actionTime) {
        if (!studentService.exists(studentId)) {
            throw new HttpEntityNotFoundException("Student with id '%d' does not exist", studentId);
        }

        studentStatusService.changeStudentStatus(StudentStatus
                .builder()
                .studentId(studentId)
                .creationTime(new Date().getTime())
                .actionTime(actionTime)
                .status(status)
                .build()
        );
    }
}

File: src/main/java/com/bdev/hogwarts_api/rest_service/student_status/StudentStatusRestServiceImpl.kt
package com.bdev.hogwarts_api.rest_service.student_status

import acropollis.municipali.security.common.dto.MunicipaliUserInfo
import com.bdev.hogwarts_api.data.dto.student.StudentStatus
import com.bdev.hogwarts_api.data.dto.student.StudentStatusType
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException
import com.bdev.hogwarts_api.service.student_legacy.StudentService
import com.bdev.hogwarts_api.service.student_status.StudentStatusService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import java.util.Date

@Service
open class StudentStatusRestServiceImpl : StudentStatusRestService {
    @Autowired
    private lateinit var studentService: StudentService
    @Autowired
    private lateinit var studentStatusService: StudentStatusService

    override fun getStudentsStatuses(userInfo: MunicipaliUserInfo): List<StudentStatus> {
        return studentStatusService.getAllStudentsStatuses()
    }

    @Transactional(readOnly = true)
    override fun getLatestStudentsStatuses(userInfo: MunicipaliUserInfo): List<StudentStatus> {
        return studentService
                .getAllStudents()
                .map { it.id ?: throw RuntimeException() }
                .mapNotNull { studentStatusService.getStudentStatus(it) }
    }

    @Transactional(readOnly = true)
    override fun getStudentStatuses(userInfo: MunicipaliUserInfo, studentId: Long): List<StudentStatus> {
        if (!studentService.exists(studentId)) {
            throw HttpEntityNotFoundException("Student with id '%d' does not exist", studentId)
        }

        return studentStatusService.getAllStudentStatuses(studentId)
    }

    @Transactional
    override fun changeStudentStatus(userInfo: MunicipaliUserInfo, studentId: Long, status: StudentStatusType, actionTime: Long) {
        if (!studentService.exists(studentId)) {
            throw HttpEntityNotFoundException("Student with id '%d' does not exist", studentId)
        }

        studentStatusService.changeStudentStatus(StudentStatus(
                studentId = studentId,
                creationTime = Date().time,
                actionTime = actionTime,
                status = status
        ))
    }
}
